### ISSUE TITLE
Audio recording from microphone

### ISSUE THREAD

**Issue by jkolobok** [bytedeco/javacv#67]
I'm trying to record a pure audio file(mp3) capturing audio from the microphone and saving it using ffmpeg recorder.
I'm getting file, I can hear sound, but there is A LOT of noise. Capturing the same way but writing it using javasound api don't produce so much sound. 
This is a small test. In the future I'll be creating video+audio from mic (demo shows the same problem - correct video and noisy audio).
Do you have any samples how to write pure audio?

**Comment by saudet** [comment]
If you're still up to creating a sample, we're still interested! Let us know what you need to get started, thanks!

**Comment by bennyveo** [comment]
Hey dudes and dudettes,

Just wanted to chime in on this thread and share a solution to capturing CLEAN audio from a TargetDataLine source (mic or other) and writing to an FFmpegFrameRecorder instance using recordSamples (similar results can be achieved using org.bytedeco.javacv.Frame and setting the samples property).

```java
// 44.1 sample rate, 16 bits, stereo, signed, little endian
AudioFormat audioFormat = new AudioFormat(44100.0F, 16, 2, true, false);

Mixer.Info[] minfoSet = AudioSystem.getMixerInfo();
		 		Mixer mixer = AudioSystem.getMixer(minfoSet[DEVICE_INDEX]);
		 		DataLine.Info dataLineInfo = new DataLine.Info( TargetDataLine.class, audioFormat );

// Open and start capturing audio    
					TargetDataLine line = (TargetDataLine) mixer.getLine(dataLineInfo);
					line.open(audioFormat);
					line.start();
					
					int audioBufferSize = (int) audioFormat.getSampleRate() * audioFormat.getChannels();
					byte[] audioBytes = new byte[audioBufferSize];
					
					AudioInputStream isS = new AudioInputStream(line);

					while (true)
					{
						try
						{
			                // Read from the line... non-blocking	
                            int nBytesRead = isS.read(audioBytes, 0, line.available());
							
							if (nBytesRead > 0)
							{
// Since we specified 16 bits in the AudioFormat, we need to convert our read byte[] to short[] (see source from FFmpegFrameRecorder.recordSamples for AV_SAMPLE_FMT_S16)
// Let's initialize our short[] array
								int nSamplesRead = nBytesRead / 2;
								short[] samples = new short[nSamplesRead];
								
// Let's wrap our short[] into a ShortBuffer and pass it to recordSamples
                             ByteBuffer.wrap(audioBytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(samples);
								ShortBuffer sBuff = ShortBuffer.wrap(samples, 0, nSamplesRead);

                                // recorder is instance of org.bytedeco.javacv.FFmpegFrameRecorder
								recorder.recordSamples((int) audioFormat.getSampleRate(), audioFormat.getChannels(), sBuff);
							}
						} 
						catch (IOException | org.bytedeco.javacv.FrameRecorder.Exception e)
						{
							e.printStackTrace();
						}
					}
```
I hope this helps anyone having issues with capture / quality :o)

**Comment by saudet** [comment]
@bennyveo Nice! Thanks a lot. Would it be possible to "bundle" that into a self-contained sample, just something like a `main()` method with your name and message above in comments, and send a pull request to have it added to the samples directory? It would be awesome!

**Comment by bennyveo** [comment]
Sure brutha, sounds good I’ll do that… Currently wrestling with getting timestamps in audio/video aligned, so let me tackle that and put it all in a single app. :o)

Cheers,

Ben Davenport
CTO | Mixify
www.mixify.com
www.facebook.com/mixify
www.twitter.com/mixify

**Comment by saudet** [comment]
New sample merged: #131 Thanks @bennyveo !
